### DISCUSSION TITLE
命令可直接运行使用os.exec os.execv 运行失败

### DISCUSSION THREAD

**Discussion by PeakRacing** [xmake-io/xmake#3946] [Q&A] [answered]
命令为 arm-none-eabi-gcc -E -P -CC -o rom_symbol.ld - < ./lib/xradio_v2/rom_symbol.ld 直接 运行正常，输出也正常，使用os.exec
os.exec(toolchains .. "/arm-none-eabi-gcc -E -P -CC "..define_flasgs.." -o rom_symbol.ld - < ./lib/xradio_v2/rom_symbol.ld")
失败 
![image](https://github.com/xmake-io/xmake/assets/24759100/399fefda-b1ea-4c64-9e54-ad2a6d904a33)
使用os.execv os.execv(toolchains .. "/arm-none-eabi-gcc",{"-E","-P","-CC",define_flasgs},{stdin = "./lib/xradio_v2/rom_symbol.ld" ,stdout = "rom_symbol.ld"})
也失败 
![image](https://github.com/xmake-io/xmake/assets/24759100/631c2977-e324-4c33-a73f-266af20a3d99)
请问应该如何修改？

**Comment by waruqi** [comment, marked as answer]
> 命令为 arm-none-eabi-gcc -E -P -CC -o rom_symbol.ld - < ./lib/xradio_v2/rom_symbol.ld 直接 运行正常，输出也正常，使用os.exec
os.exec(toolchains .. "/arm-none-eabi-gcc -E -P -CC "..define_flasgs.." -o rom_symbol.ld - < ./lib/xradio_v2/rom_symbol.ld")
失败

os.exec 执行的是进程，不是 shell，不支持 shell 的写法。。用 os.execv

> 使用os.execv os.execv(toolchains .. "/arm-none-eabi-gcc",{"-E","-P","-CC",define_flasgs},{stdin = "./lib/xradio_v2/rom_symbol.ld" ,stdout = "rom_symbol.ld"})

1. 传入 stdin/stdout 的路径是相对路径，确认当前目录是否匹配。。
2. `-o rom_symbol.ld` 是传参，不是重定向，不应该放到 stdout 
3. 最后面那个 `-` 参数也漏了

**Comment by PeakRacing** [comment]
路径是对的 
![image](https://github.com/xmake-io/xmake/assets/24759100/3824dd14-2491-44ab-9f1d-2d648febdb19)
 ，修改为 os.execv(toolchains .. "/arm-none-eabi-gcc",{"-E","-P","-CC",define_flasgs,"-o","./rom_symbol.ld"},{stdin = "./lib/xradio_v2/rom_symbol.ld"}) 依旧 
![image](https://github.com/xmake-io/xmake/assets/24759100/04323171-d04e-4ffe-967d-ac8d0a63de2a)

**Reply by PeakRacing** [reply]
@waruqi

**Reply by waruqi** [reply]
> 最后面那个 - 参数也漏了

**Reply by PeakRacing** [reply]
- 参数是哪里？没看出来，大佬指一下

**Reply by waruqi** [reply]
```
{"-E","-P","-CC",define_flasgs,"-o","./rom_symbol.ld", "-"}
```

> -o rom_symbol.ld - < ./lib/xradio_v2/rom_symbol.ld")

我看你命令行里不是还有个 `-` 么

**Reply by PeakRacing** [reply]
是我瞎了，哈哈哈哈。没往后看光盯着前面了，可以了
